tools: Add a Rust-based standard OCI container runtime based on Kata agent

ManaSugi opened the pull request:
Add a Rust-based standard OCI container runtime based on Kata agent.

Fixes: #2784

**Below is the README.**

### Change log

**Apr/28th/2022:**  Update to the latest README
**Nov/5th/2021:**  Mark seccomp support in the table of features compared to `runc`
**Oct/23th/2021:** Add `spec` command to the current status table
**Oct/13th/2021:** Add current status of `runk` compared to `runc`

# runk

## Overview

> **Warnings:**
> `runk` is currently an experimental tool.
> Only continue if you are using a non-critical system.

`runk` is a standard OCI container runtime written in Rust based on a modified version of
the [Kata Container agent](https://github.com/kata-containers/kata-containers/tree/main/src/agent), `kata-agent`.

`runk` conforms to the [OCI Container Runtime specifications](https://github.com/opencontainers/runtime-spec).

Unlike the [Kata Container runtime](https://github.com/kata-containers/kata-containers/tree/main/src/agent#features),
`kata-runtime`, `runk` spawns and runs containers on the host machine directly.
The user can run `runk` in the same way as the existing container runtimes such as `runc`,
the most used implementation of the OCI runtime specs.

## Why does `runk` exist?

The `kata-agent` is a process running inside a virtual machine (VM) as a supervisor for managing containers
and processes running within those containers.
In other words, the `kata-agent` is a kind of "low-level" container runtime inside VM because the agent
spawns and runs containers according to the OCI runtime specs.
However, the `kata-agent` does not have the OCI Command-Line Interface (CLI) that is defined in the
[runtime spec](https://github.com/opencontainers/runtime-spec/blob/master/runtime.md).
The `kata-runtime` provides the CLI part of the Kata Containers runtime component,
but the `kata-runtime` is a container runtime for creating hardware-virtualized containers running on the host.

`runk` is a Rust-based standard OCI container runtime that manages normal containers,
not hardware-virtualized containers.
`runk` aims to become one of the alternatives to existing OCI compliant container runtimes.
The `kata-agent` has most of the [features](https://github.com/kata-containers/kata-containers/tree/main/src/agent#features)
needed for the container runtime and delivers high performance with a low memory footprint owing to the
implementation by Rust language.
Therefore, `runk` leverages the mechanism of the `kata-agent` to avoid reinventing the wheel.

## Performance

`runk` is faster than `runc` and has a lower memory footprint.

This table shows the average of the elapsed time and the memory footprint (maximum resident set size)
for running sequentially 100 containers, the containers run `/bin/true` using `run` command with
[detached mode](https://github.com/opencontainers/runc/blob/master/docs/terminals.md#detached)
on 12 CPU cores (`3.8 GHz AMD Ryzen 9 3900X`) and 32 GiB of RAM.
`runk` always runs containers with detached mode currently.

Evaluation Results:

|                       | `runk` (v0.0.1) | `runc` (v1.0.3) | `crun` (v1.4.2) |
|-----------------------|---------------|---------------|---------------|
| time [ms]           | 39.83         | 50.39         | 38.41         |
| memory footprint [MB] | 4.013         | 10.78         | 1.738         |

## Status of `runk`

We drafted the initial code here, and any contributions to `runk` and [`kata-agent`](https://github.com/kata-containers/kata-containers/tree/main/src/agent)
are welcome.

Regarding features compared to `runc`, see the `Status of runk` section in the [issue](https://github.com/kata-containers/kata-containers/issues/2784).

## Building

`runk` uses the modified the `kata-agent` binary, `oci-kata-agent`, which is an agent to be called from `runk`.
Therefore, you also need to build the `oci-kata-agent` to run `runk`.

You can build both `runk` and `oci-kata-agent` as follows.

```bash
$ cd runk
$ make
```

To install `runk` and `oci-kata-agent` into default directory for install executable program (`/usr/local/bin`):

```bash
$ sudo make install
```

## Using `runk` directly

Please note that `runk` is a low level tool not developed with an end user in mind.
It is mostly employed by other higher-level container software like `containerd`.

If you still want to use `runk` directly, here's how.

### Prerequisites

It is necessary to create an OCI bundle to use the tool. The simplest method is:

``` bash
$ bundle_dir="bundle"
$ rootfs_dir="$bundle_dir/rootfs"
$ image="busybox"
$ mkdir -p "$rootfs_dir" && (cd "$bundle_dir" && runk spec)
$ sudo docker export $(sudo docker create "$image") | tar -C "$rootfs_dir" -xvf -
```

> **Note:**
> If you use the unmodified `runk spec` template, this should give a `sh` session inside the container.
> However, if you use `runk` directly and run a container with the unmodified template,
> `runk` cannot launch the `sh` session because `runk` does not support terminal handling yet.
> You need to edit the process field in the `config.json` should look like this below
> with `"terminal": false` and `"args": ["sleep", "10"]`.

```json
"process": {
    "terminal": false,
    "user": {
        "uid": 0,
        "gid": 0
    },
    "args": [
        "sleep",
        "10"
    ],
    "env": [
        "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "TERM=xterm"
    ],
    "cwd": "/",
    [...]
}
```

If you want to launch the `sh` session inside the container, you need to run `runk` from `containerd`.

Please refer to the [Using `runk` from containerd](#using-runk-from-containerd) section

### Running a container

Now you can go through the [lifecycle operations](https://github.com/opencontainers/runtime-spec/blob/master/runtime.md)
in your shell.
You need to run `runk` as `root` because `runk` does not have the rootless feature which is the ability
to run containers without root privileges.

```bash
$ cd $bundle_dir

# Create a container
$ sudo runk create test

# View the container is created and in the "created" state
$ sudo runk state test

# Start the process inside the container
$ sudo runk start test

# After 10 seconds view that the container has exited and is now in the "stopped" state
$ sudo runk state test

# Now delete the container
$ sudo runk delete test
```

## Using `runk` from `containerd`

`runk` can run containers with the containerd runtime handler support on `containerd`.

### Prerequisites for `runk` with containerd

* `containerd` v1.2.4 or above
* `cri-tool`

> **Note:**
> [`cri-tools`](https://github.com/kubernetes-sigs/cri-tools) is a set of tools for CRI
> used for development and testing.

Install `cri-tools` from source code:

```bash
$ go get github.com/kubernetes-incubator/cri-tools
$ pushd $GOPATH/src/github.com/kubernetes-incubator/cri-tools
$ make
$ sudo -E make install
$ popd
```

Write the `crictl` configuration file:

``` bash
$ cat <<EOF | sudo tee /etc/crictl.yaml
runtime-endpoint: unix:///run/containerd/containerd.sock
EOF
```

### Configure `containerd` to use `runk`

Update `/etc/containerd/config.toml`:

```bash
$ cat <<EOF | sudo tee /etc/containerd/config.toml
version = 2
[plugins."io.containerd.runtime.v1.linux"]
  shim_debug = true
[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
  runtime_type = "io.containerd.runc.v2"
[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runk]
  runtime_type = "io.containerd.runc.v2"
EOF
```

Restart `containerd`:

```bash
$ sudo systemctl restart containerd
```

### Running a container with `crictl` command line

You can run containers in `runk` via containerd's CRI.

Pull the `busybox` image:

``` bash
$ sudo crictl pull busybox
```

Create the sandbox configuration:

``` bash
$ cat <<EOF | tee sandbox.json
{
    "metadata": {
        "name": "busybox-sandbox",
        "namespace": "default",
        "attempt": 1,
        "uid": "hdishd83djaidwnduwk28bcsb"
    },
    "log_directory": "/tmp",
    "linux": {
    }
}
EOF
```

Create the container configuration:

``` bash
$ cat <<EOF | tee container.json
{
    "metadata": {
        "name": "busybox"
    },
    "image": {
        "image": "docker.io/busybox"
    },
    "command": [
        "sh"
    ],
    "envs": [
        {
            "key": "PATH",
            "value": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        },
        {
            "key": "TERM",
            "value": "xterm"
        }
    ],
    "log_path": "busybox.0.log",
    "stdin": true,
    "stdin_once": true,
    "tty": true
}
EOF
```

With the `crictl` command line of `cri-tools`, you can specify runtime class with `-r` or `--runtime` flag.

Launch a sandbox and container using the `crictl`:

```bash
# Run a container inside a sandbox
$ sudo crictl run -r runk container.json sandbox.json
f492eee753887ba3dfbba9022028975380739aba1269df431d097b73b23c3871

# Attach to the running container
$ sudo crictl attach --stdin --tty f492eee753887ba3dfbba9022028975380739aba1269df431d097b73b23c3871
/ #
```

ManaSugi commented:
I would like to discuss if the `runk` is of value to Kata community. 
Thank you.

sameo commented:
Hey @ManaSugi, thanks for the PR. This is very exciting and it's nice to see `kata-agent` being used that way.

I guess my first question would be: How would you compare `runk` against [`youki`](https://github.com/containers/youki) which is arguably the leading rust based OCI runtime?

devimc commented:
@ManaSugi thanks for raising this, would you like to  discuss this in the following Kata Architecture Commitee meeting ? 
https://etherpad.opendev.org/p/Kata_Containers_2021_Architecture_Committee_Mtgs

ManaSugi commented:
Hello @sameo 
Thank you for your comment and question.
I think that the features for spawning and running containers that `kata-agent` and `youki` have are not so different.
`youki` OCI features: https://github.com/containers/youki#status-of-youki
`kata-agent` OCI features: https://github.com/kata-containers/kata-containers/tree/main/src/agent#features
(`kata-agent` is ready to support Seccomp feature #1788.)

So, if `runk` supports more CLI command operations such as `exec`, `ps`, etc. and we can prepare more tests, `runk` has a good chance of becoming a leading rust-based OCI runtime. Even if `runk` could not become the leading runtime, I guess that the number of users who use `kata-agent` through `runk` will increase. As a result, `runk` will lead to boost the development speed of `kata-agent`.

The only thing is that `runk` need to fork/exec `kata-agent` and it causes an additional overhead, whereas `youki` is one binary.
We can avoid the overhead by calling APIs inside `rustjail` directly. However, it is hard to achieve this without significantly changing the existing `kata-agent` code since `rustjail` has been developed on the assumption that it is called from `kata-agent`.

Hello @devimc 
Thank you for inviting me to the meeting. I would like to discuss `runk` there. Could you tell me if there is something I have to do to attend the meeting?

devimc commented:
@ManaSugi add a new topic to the agent https://etherpad.opendev.org/p/Kata_Containers_2021_Architecture_Committee_Mtgs and join the meeting
Dial-In link: https://zoom.us/j/91444082924
PASSCODE: "kata"

ManaSugi commented:
@devimc I added a topic to the agenda.

dgibson commented:
Honestly, implementing an OCI runtime in terms of kata-agent, rather than just rustjail seems like an odd choice to me.  Most of what the kata agent (other than rustjail) deals with is essentially kata specific - how to translate devices, networks and mounts from outside the VM into the VM.  Likewise a lot of the information sent over the grpc link to the agent also only makes sense in the context of a VM.

ManaSugi commented:
/test

c3d commented on tools/runk/README.md:
```diff
@@ -0,0 +1,265 @@
+
+The `runk` is faster than `runc` and has a lower memory footprint.
+
+This table shows the average of the elapsed time and the memory footprint for running sequentially 500 containers,
```
@ManaSugi Could you clarify if the numbers in the table below are for `runk` by itself, or for `runk` + the `oci-kata-agent`?

ManaSugi replied:
@c3d Thank you for the comment. The table shows `runk` + `oci-kata-agent` because `runk` calls `oci-kata-agent` internally for spawning and running a container.

ManaSugi commented:
Resolved conflicts and updated `agent/Makefile` and `runk/Cargo.lock`

ManaSugi commented:
I would appreciate it if anyone could review `runk`. If there is no problem, I'll squash the commits.

jodh-intel commented on tools/runk/src/main.rs:
```diff
@@ -0,0 +1,241 @@
+    let console_socket = Arg::with_name("console-socket")
+        .long("console-socket")
+        .takes_value(true)
+        .help("path to an AF_UNIX socket which will receive a file descriptor referencing the master end of the console's pseudoterminal");
```
Nit: "pseudo-terminal".

ManaSugi replied:
I fixed it.

jodh-intel commented on tools/runk/src/main.rs:
```diff
@@ -0,0 +1,241 @@
+        .long("no-pivot")
+        .help("do not use pivot root to jail process inside rootfs");
+
+    let matches = App::new("runk")
+        .version("0.1.0")
+        .author("The Kata Containers community <kata-dev@lists.katacontainers.io>")
```
Can you use clap's `crate_name!()`, `crate_version!()` and `crate_authors!()` macros here instead of hard-coding the values?

ManaSugi replied:
I fixed it.

jodh-intel commented on tools/runk/src/main.rs:
```diff
@@ -0,0 +1,231 @@
+        logger = slog::Logger::root(slog::Discard, o!());
+    }
+
+    let state_root = Path::new(matches.value_of("root").unwrap());
```
Yes, but the problem is you have two pieces of code which guarantee state, but they are disconnected (and thus likely to deviate over time). We're trying to remove as many `unwrap()` and `expect()` calls as practical and this can be avoided with a slight reworking of the code.

ManaSugi replied:
I understand. I removed all `unwrap()` except for test.

jodh-intel commented on tools/runk/src/spec.rs:
```diff
@@ -0,0 +1,206 @@
+
+fn generate_config() -> String {
+    r#"{
+	"ociVersion": "1.0.2-dev",
```
Might be worth making this a `const`.

ManaSugi replied:
I fixed it.

jodh-intel commented on tools/runk/src/start.rs:
```diff
@@ -0,0 +1,46 @@
+    let mut status = Status::load(state_root, id)?;
+    if state::get_current_container_state(&status)? != ContainerState::Created {
+        return Err(anyhow!(
+            "cannot start a container that is not in created status"
```
Why not log the current statu to make the error clearer?

ManaSugi replied:
I fixed it.

jodh-intel commented on tools/runk/src/state.rs:
```diff
@@ -0,0 +1,104 @@
+    let mut has_fifo = false;
+
+    if running {
+        let fifo = format!("{}/{}/{}", status.bundle, status.id, EXEC_FIFO_FILENAME);
```
What if `status.id == ""`?

ManaSugi replied:
When launching container using `runk`, it is mandatory to input the `container id` (empty string is prohibited) in the CLI, so I think that `status.id` won't be empty.

jodh-intel reviewed (commented):
Thanks @ManaSugi. I've done a quick review but will look at this more closely on Monday.

ManaSugi commented:
@jodh-intel I appreciate for the review and I'll fix them. 
In addition, I'll modify CLI component of `runk` using `clap` [Derive Macro](https://github.com/clap-rs/clap#using-derive-macros) for the future use cases as mentioned in [this comment](https://github.com/kata-containers/kata-containers/issues/2784#issuecomment-969200057).

jodh-intel commented:
@ManaSugi - are you still planning to update this PR?

ManaSugi commented:
@jodh-intel I apologize for the late update. I'll update this PR next week.

jodh-intel commented:
Hi @ManaSugi - any update on this?

jodh-intel commented:
PR marked as stale.

ManaSugi commented:
@jodh-intel  Sorry for the late update, I'll update this using [`liboci-cli`](https://github.com/containers/youki/tree/main/crates/liboci-cli) as command line interface of `runk` for improving maintainability by the end of the week. In addtion, I'll squash the commits at the same time.

ManaSugi commented on src/agent/rustjail/src/container.rs:
```diff
@@ -663,10 +673,14 @@ fn do_init_child(cwfd: RawFd) -> Result<()> {
+        #[cfg(feature = "standard-oci-runtime")]
+        if csocket_fd.is_some() {
+            console::setup_master_console(csocket_fd.unwrap())?;
+        }
```
Thank you for the review.
I removed all `unwrap()` method except for test.

ManaSugi commented on tools/runk/Cargo.toml:
```diff
@@ -0,0 +1,35 @@
+slog-stdlog = "4.1.0"
+slog-async = "2.3.0"
+async-trait = "0.1.31"
+futures = "0.3"
```
I fixied it.

ManaSugi commented on tools/runk/Cargo.toml:
```diff
@@ -0,0 +1,35 @@
+tokio = { version = "1.2.0", features = ["full"] }
+serde = { version = "1.0.127", features = ["derive"] }
+serde_json = "1.0.66"
+libc = "0.2"
```
I fixed it.

ManaSugi commented on tools/runk/README.md:
```diff
@@ -0,0 +1,306 @@
+
+This table shows the average of the elapsed time and the memory footprint (maximum resident set size)
+for running sequentially 500 containers, the containers run `/bin/true` with `detached` mode; `runk`
+always runs containers with the mode currently:
```
I fixed it.

ManaSugi commented on tools/runk/README.md:
```diff
@@ -0,0 +1,306 @@
+However, if you use `runk` directly and run a container with the unmodified template,
+`runk` cannot launch the `sh` session because `runk` does not support terminal handling yet.
+You need to edit the process field in the `config.json` should look like this below
+with `"terminal": false` and `"args": ["sleep", "10"]`.
```
I fixed it.

ManaSugi commented on tools/runk/src/cgroup.rs:
```diff
@@ -0,0 +1,56 @@
+
+// Remove cgroup path recursively by removing the subdirectories (sub-cgroups).
+// std::fs::remove_dir_all can not be used to remove the cgroup path because
+// the function also tries to remove files in the directories and fails.
```
Thank you for the review.
I removed my`remove_dir` function.
Regarding `sleep`, I use it for a conservative approach to delete cgroups path. If container processes exist in the cgroups, `runk` cannot delete the cgroups directory due to a `Device or resource busy` error. Before calling `remove_cgroup_dir`, `runk` kills the container process, but it may take some time for the process to finish. Hence, `runk` uses `sleep` approach to wait for the process to finish without an infinite loop.
`runc` also takes the same approach:
https://github.com/opencontainers/runc/blob/6ff042023ce8f550a70d3c0bd2588e48df70d939/libcontainer/cgroups/utils.go#L270

ManaSugi commented on tools/runk/src/container.rs:
```diff
@@ -0,0 +1,158 @@
+
+impl ContainerContext {
+    pub fn new(context: &ArgMatches, config: &str, state_root: &Path) -> Result<Self> {
+        let bundle = Path::new(context.value_of("bundle").unwrap()).canonicalize()?;
```
I fixed it.

ManaSugi commented:
@jodh-intel I apoplogize for the late update.
I redesigned `runk` using [`liboci-cli`](https://github.com/containers/youki/tree/main/crates/liboci-cli) as the cli of `runk` and cleaned the commits.

utam0k commented on src/tools/runk/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+rustjail = { path = "../../agent/rustjail", features = ["standard-oci-runtime"] }
+oci = { path = "../../libs/oci" }
+logging = { path = "../../libs/logging" }
+liboci-cli = { git = "https://github.com/containers/youki.git" }
```
Hi! I am a maintainer of youki. I have the plan to publish `liboci-cli` to crates.io soon.
https://github.com/containers/youki/issues/586

ManaSugi replied:
Thank you for the information. It's awesome!

jodh-intel replied:
Any update on publishing on crates.io @utam0k?

fidencio commented:
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

jodh-intel commented:
Any update on this @ManaSugi? Is it ready for re-review?

ManaSugi commented:
@jodh-intel I have no updates. If you don't mind, could you re-review it?. Thank you.

jodh-intel commented on src/tools/runk/README.md:
```diff
@@ -0,0 +1,277 @@
+
+The `kata-agent` is a process running inside a virtual machine (VM) as a supervisor for managing containers
+and processes running within those containers.
+In other words, the `kata-agent` is a kind of "low-level" container runtime inside VM because the agent spawns and runs containers according to the OCI runtime specs.
```
Nit: Overly long line.

ManaSugi replied:
I fixed it.

jodh-intel commented on src/tools/runk/README.md:
```diff
@@ -0,0 +1,277 @@
+
+## Performance
+
+`runk` is faster than `runc` and has a lower memory footprint.
+
+This table shows the average of the elapsed time and the memory footprint (maximum resident set size)
+for running sequentially 500 containers, the containers run `/bin/true`
+with [detached mode](https://github.com/opencontainers/runc/blob/master/docs/terminals.md#detached) `runk`
+always runs containers with detached mode currently:
+
+|                       | runk (+ oci-kata-agent) | runc  | crun  |
+|-----------------------|-------------------------|-------|-------|
+| time [msec]           | 39.01                   | 53.09 | 35.35 |
+| memory footprint [MB] | 4.223                   | 16.54 | 3.157 |
```
It might be worth adding version details, architectures and hardware tested on here.

ManaSugi replied:
I fixed it.

jodh-intel commented on src/tools/runk/README.md:
```diff
@@ -0,0 +1,277 @@
+
+## Status of runk
+
+`runk` is currently an experimental tool.
```
I would consider moving this section to the top of the file and making it a ["warning"](https://github.com/kata-containers/kata-containers/blob/main/docs/Documentation-Requirements.md#warnings-and-other-admonitions). Something like:

> **Warning:** `runk` is currently an experimental tool.

ManaSugi replied:
I fixed it.

jodh-intel commented on src/tools/runk/README.md:
```diff
@@ -0,0 +1,277 @@
+
+## Building
+
+> **Warnings:**
+> The runk is experimental.
+> Only continue if you are using a non-critical system.
```
I'd add that text into the warning at the top of the file.

ManaSugi replied:
I fixed it.

jodh-intel commented on src/tools/runk/README.md:
```diff
@@ -0,0 +1,277 @@
+$ cd $bundle_dir
+
+# Create a container
+$ sudo runk create test
```
I think you should explain why it is necessary to run `runk` as `root`.

ManaSugi replied:
I fixed it.

jodh-intel reviewed (commented):
Thanks @ManaSugi - just a few doc comments...

jodh-intel commented:
@ManaSugi - one other thing, if and when this lands, how is it going to be tested? Do you plan to create some integration tests in https://github.com/kata-containers/tests? If so, you might want to speak to @GabyCT and @dborquez as they can probably help. We removed a lot of the old Docker tests when we moved to the [shimv2 architecture](https://github.com/kata-containers/kata-containers/tree/main/docs/design/architecture#shim-v2-architecture) but these could be re-used for `runk` possibly.

utam0k commented:
@jodh-intel 
I've been a little busy with my personal life and haven't had much time, but I have a plan. Can you please wait a little longer?
https://github.com/kata-containers/kata-containers/pull/2785#discussion_r807816694

jodh-intel commented:
Hi @utam0k - Sure! I was just wondering :smile:

ManaSugi commented:
@jodh-intel Thank you for your review. I fixed them.

> @ManaSugi - one other thing, if and when this lands, how is it going to be tested? Do you plan to create some integration tests in https://github.com/kata-containers/tests? If so, you might want to speak to @GabyCT and @dborquez as they can probably help. We removed a lot of the old Docker tests when we moved to the [shimv2 architecture](https://github.com/kata-containers/kata-containers/tree/main/docs/design/architecture#shim-v2-architecture) but these could be re-used for `runk` possibly.

Yes, I plan to create integration tests in https://github.com/kata-containers/tests for runk. Currently, runk cannot run with Docker, but when runk starts support for Docker, I'll try to re-use the old Docker tests. Thank you.

jodh-intel commented:
/test

jodh-intel reviewed (approved):
Thansk @ManaSugi.

lgtm

jodh-intel commented:
Ping for reviews - this PR has been open a very long time so would be good to land it folks...

jodh-intel commented on src/tools/runk/README.md:
```diff
@@ -0,0 +1,283 @@
+
+`runk` can run containers with the containerd runtime handler support on `containerd`.
+
+### Prerequisites
```
Hi @ManaSugi - this is causing the CI to fail since there are two identical heading names in the same document. You could change this to "### Prerequisites for runk with containerd" or similar?

ManaSugi replied:
Thank you for the suggestion. I fixed it.

jodh-intel commented:
Can we get another review over here folks? This is a big PR and would be great to see this land. It already has one ack, so we need one more ack to land this new runtime!

@liubin, @bergwolf, @lifupan - could one of you tal?

utam0k commented:
Youki team has released the new version, and `liboci-cli` appeared on crates.io.
https://crates.io/crates/liboci-cli

arronwy commented:
Can one of the admins verify this patch?

ManaSugi commented:
I solved the conflicts and added 4 commits.

zvonkok commented:
@ManaSugi Does `runk` support OCI runtime spec hooks?

ManaSugi commented:
@zvonkok Yes, `runk` can use OCI hooks since `kata-agent` already supports the hooks, `prestart`, `poststart` and `poststop`.
(`prestart` is deprecated currently, so I have plan to add `createRuntime`, `createContainer`, `startContainer` to the agent instead.)

ManaSugi commented:
/test

jodh-intel commented:
Please can someone else tal at this PR?

fidencio commented:
/retest-devmapper-clh
/retest-qat
/retest-ubuntu-metrics

fidencio commented:
/retest-ubuntu-qat

fidencio reviewed (approved):
Let's have this one in for the 2.5.0-alpha1.

Thanks @ManaSugi!

ManaSugi commented:
I appreciate for the review. If there is no problem, I'd like to squash the commits into initial two commits (e30ca624c7b60b85ccd99e96b7e4821615aa96e4 and e02c0ac60763393b24627f1531c981a9c1b17d2c).

jodh-intel commented:
@ManaSugi - wfm, thanks :smile: I've added the dnm label to allow you to do that.

ManaSugi commented:
@jodh-intel Thank you for adding the label. I squashed the commits.

jodh-intel commented:
/test

ManaSugi commented:
/retest-devmapper-clh
/retest-ubuntu

liubin commented on src/agent/rustjail/src/container.rs:
```diff
@@ -379,6 +392,9 @@ fn do_init_child(cwfd: RawFd) -> Result<()> {
     let cm: FsManager = serde_json::from_str(cm_str)?;
 
+    #[cfg(feature = "standard-oci-runtime")]
+    let csocket_fd = console::setup_console_socket(&std::env::var(CONSOLE_SOCKET_FD)?)?;
```
Can this line be moved to before line 692?

liubin commented on src/tools/runk/README.md:
```diff
@@ -0,0 +1,282 @@
+[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
+  runtime_type = "io.containerd.runc.v2"
+[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runk]
+  runtime_type = "io.containerd.runc.v2"
```
`io.containerd.runc.v2` or `io.containerd.runk.v2` ?

liubin commented on src/tools/runk/src/main.rs:
```diff
@@ -0,0 +1,111 @@
+        exit(1);
+    }
+
+    exit(0);
```
Is this required?

jodh-intel commented:
Thanks @ManaSugi !

liubin commented on src/tools/runk/libcontainer/src/cgroup.rs:
```diff
@@ -0,0 +1,40 @@
+        retries -= 1;
+    }
+
+    return Err(anyhow!("failed to remove cgroups paths: {:?}", path));
```
Nit: paths -> path

jodh-intel commented:
Hi @liubin - apologies, I hadn't seen your comments before I hit the big green button :smile:

@ManaSugi - Could you possibly raise a follow-up PR to address @liubin's comments?
